src/panic.md

```diff
@@ -0,0 +1,102 @@
+configured (via `panic=abort`) to always be non-recoverable. (The converse is
+not true: `panic=unwind` does not guarantee that all panics are recoverable,
+only that panicking via the `panic!` macro and similar standard library
+mechanisms is recoverable.) When panic recovery occurs, the runtime "unwinds"
+Rust frames, just as C++'s `throw` unwinds C++ frames, until the panic reaches
+the point of recovery (for instance at a thread boundary). This means that as
+the panic traverses Rust frames, live objects in those frames that [implement
+`Drop`][destructors] will have their `drop` methods called. Thus, when normal
+execution resumes, no-longer-accessible objects will have been "cleaned up"
+just as if they had gone out of scope normally.
+
+> **Note**: As long as this guarantee of resource-cleanup is preserved,
+> "unwinding" may be implemented without actually using the mechanism used by
+> C++ for the target platform.
+
+> **Note**: The standard library provides two mechanisms for recovering from a panic,
+> [`catch_unwind`][fn-catch-unwind] (which enables recovery within the
+> panicking thread) and [`thread::spawn`][thread-spawn] (which automatically
+> sets up panic recovery for the spawned thread so that other threads may
+> continue running).
+
+### Unwinding across FFI boundaries
+[unwinding-ffi]: #unwinding-across-ffi-boundaries
+
+It is possible to unwind across FFI boundaries using an [appropriate ABI
+declaration][unwind-abi]. While useful in certain cases, this creates unique
+opportunities for undefined behavior, especially when multiple language runtimes
+are involved.
+
+Unwinding with the wrong ABI is undefined behavior:
+
+* Causing an unwind into Rust code from a foreign function that was called via a
+  function declaration or pointer declared with a non-unwinding ABI, such as `"C"`,
+  `"system"`, etc. (For example, this case occurs when such a function written in
+  C++ throws an exception that is uncaught and propagates to Rust.)
+* Calling a Rust `extern` function that unwinds (with `extern "C-unwind"` or
+  another ABI that permits unwinding) from a runtime that does not support.
+  unwinding, such as code compiled with GCC or Clang using `-fno-exceptions`
+
+Catching a foreign unwinding operation (such as a C++ exception) using
+`catch_unwind`, `JoinHandle::join`, or by letting it propagate all the way to a
+Rust `main()` function will have one of two behaviors, and it is unspecified
+which will occur:
+* The process aborts.
+* The function returns a `Result::Err` containing an opaque type.
+
+Note that Rust code compiled or linked with a different runtime counts as a
+"foreign exception" for the purpose of this guarantee. Thus, a library that
+uses `panic!` and is linked against one version of the Rust standard library,
+invoked from an application that uses a different version of the standard
+library, may cause the entire application to crash even if the library is only
+used within a child thread.
```

Comment:
How do we even distinguish the two Rust runtimes? Isn't that entirely defined by the `exception_class`, which we hard-code to `0x4d4f5a_00_52555354`? So a panic from a different runtime will look exactly like a panic we threw ourselves? Doesn't that cause problems with the panic count?

Reply:
@nbdd0121, @bjorn3 ?

Reply:
(And I think @Amanieu or maybe @workingjubilee ?)

Reply:
What you say is correct. In order to correctly handle this, we probably would have to make the exact string we look for be dependent on the version or some other build time constant.

Reply:
What can go wrong, if the panic implementation doesn't actually change between the versions in question? Is this only a problem if we actually change the panic implementation in the future, e.g. by removing the panic-count?

Reply:
Hm, no, it's more that the panic count is a problem, I think?

- Rust A initiates `catch_unwind` and calls into Rust B.
- Rust B panics.
- Rust A catches it, then decreases the panic count...
- ...but Rust A had 0 panics, so it's now `usize::MAX` which means infinite panics and the `ALWAYS_ABORT_FLAG` has been set.

Reply:
Obviously we could modify a significant amount of internal logic to handle this case but doing all that would be pretty wild.

Reply:
😬😖

Reply:
I am pretty sure this is intensely platform-specific in reality, but if I understand everything correctly, there is a strong argument for trying to, as much as possible, pretend all exceptions are potentially foreign, and handle them as equally as possible. Yes we do have the unwind-vs-not distinction, but we probably want to try to avoid privileging Rust code differently than C-unwind.

Reply:
So what can we do to identify "foreign" Rust exceptions, since they'll have the same hardcoded exception class? Pseudo-randomize the exception class at link time?

Reply:
hmm, I'm not entirely sure, having given it some thought, since the answer relates partly to linkage I think.

Reply:
Can't we "just" make the process abort when the panic count underflows from 0 to usize::MAX?

Reply:
I *suppose*, yeah, but I'm not confident that's the only possible issue, it was just an evil one I thought of off the top of my head?

Reply:
Oh cool, I counseled with @bjorn3 because I wasn't quite satisfied with my understanding, And everything's a bit weirder than I thought:
- If C++ *does* catch our panic as an exception, we won't ever decrement the panic count. So now we're locked into "panic is abort" mode! Cool!
- Rust catching Rust is actually possible to identify correctly as a foreign version because we check equality of a pointer to this static, and two instances of CANARY should only ever unify if both code instances were statically linked together: https://github.com/rust-lang/rust/blob/5bce6d48ff09dcb2613278ec93013795718478ef/library/panic_unwind/src/gcc.rs#L45-L48

Reply:
I'm really annoyed that there's code in at least three different crates to implement our unwinding strategy and no one ever has sat down and written a high-level documentation of it!

Reply:
> If C++ does catch our panic as an exception, we won't ever decrement the panic count. So now we're locked into "panic is abort" mode! Cool!

Where does that happen?
I thought it would call our `unwind_cleanup` which aborts?

Reply:
do you mean our `exception_cleanup`?

Reply:
I do expect that is what will typically happen, but there are a lot of "should"s in the Itanium C++ Exception Handling ABI that prefix every statement which suggests such, and I'm not sure how to interpret them.

Reply:
So what happens if that does not get called? You said we're getting into "always abort on panic" mode -- and that's a good thing?

Reply:
Better than it being UB, that's for sure.

Reply:
It seems to me that we've wandered away from the original question a bit here. The original question was, since we're declaring that Rust runtimes can be "foreign", how can the implementation even distinguish a "foreign" panic from a "native" panic?

I see two possible resolutions to this, but please correct me if there are more:

* Implement a way for Rust to distinguish between panics thrown from two different Rust runtimes operating within a single process space. (Maybe every panic object could contain a pointer to a global static singleton that's part of the `core` library, and if the pointer doesn't have the expected value, we know the runtime was "foreign"?)
* As @workingjubilee mentioned above, somehow dissolve the difference between "foreign exceptions" and "Rust exceptions." I'm not sure how this would work, exactly; possibly we'd need to get rid of the panic-count concept entirely?

Reply:
> Implement a way for Rust to distinguish between panics thrown from two different Rust runtimes operating within a single process space. (Maybe every panic object could contain a pointer to a global static singleton that's part of the core library, and if the pointer doesn't have the expected value, we know the runtime was "foreign"?)

so, to reiterate, the answer is that I was slightly confused and we *do* do this on systems where we use the modal Itanium C++ ABI (as interpreted by GCC, anyways).

Reply:
Well that is a relief! Sorry, I understand the bullet clarifying that now. So @RalfJung can this thread be marked resolved?

Reply:
That was this global static you pointed at, right? [This one](https://github.com/rust-lang/rust/blob/125b26acf6f3127ecdd344c372691cefe0e9243e/library/panic_unwind/src/gcc.rs#L48)?

Seems like if we catch a mismatch there, we call `__rust_foreign_exception` which aborts the process. Yeah sounds good, thanks for finding this @workingjubilee !

Reply:
Turns out we even have a test for this. :)
Added in https://github.com/rust-lang/rust/pull/102721.